### ISSUE TITLE
SLEIGH: Instruction behaviour dependent on a STATUS register bit

### ISSUE THREAD

**Issue by philpem** [NationalSecurityAgency/ghidra#5007]
Another question open to SLEIGH experts ... this is a follow on from #5002 

The MELPS-740 introduces a "T" flag. This changes the behaviour of the ADC, SBC, AND, ORA and EOR instructions. Instead of operating on the accumulator (A register), they operate on the zero-page memory location pointed to by the X register. To express that in SLEIGH --

With T=0 the 740 behaves like a 6502:
```
:ADC OP1     is (cc=1 & aaa=3 & T=0) ... & OP1
{
	local op1 = OP1;
	local tmpC = C;
	
	C = carry(A, op1);
	
	A = A + op1 + tmpC;

	resultFlags(A);
	V = C;
}
```

With T=1:
```
:ADC OP1     is (cc=1 & aaa=3 & T=1) ... & X & OP1
{
	local op1 = OP1;
	local tmpC = C;
	local tmpX = *:1 X;
	
	C = carry(tmpX, op1);
	
	tmpX = tmpX + op1 + tmpC;

	resultFlags(tmpX);
	V = C;
}
```

Unfortunately this syntax doesn't work; SLEIGH complains that "T" isn't a varnode.

I've also tried combining the two SLEIGH descriptions into one, and adding an `if` conditional in the code -- this got me a syntax error (SLEIGH was expecting a 'goto' statement after the 'if (T==0)'). I guess this could work if I used pcode-relative branching, but that seems like a bit of an icky solution.

A context variable seems like it could be what I need, but I'd somehow have to keep the T status flag and the context variable in sync - the T flag will be pushed onto and popped off of the stack as part of the status byte.

**Comment by emteere** [comment]
You might be able to get away with a choosery statement something like:

local tmpX = (zext(T == 0) * A) + (zext(T == 1) * (*:1 X));

You could also use if stmts in the pcode, but as you say, icky.

The down fall of the choosery is the memory in page 0 would always be accessed to get a value and you may end up with a reference created to it even if the T is set to 0.  Although that may occur with internal goto's as well.

The decompiler will see the value of T as a constant and simplify away one of the cases.

In either case you should set T to an initial value, either 0 or 1 based on which is more normal in the .pspec file.

**Comment by Danil6969** [comment]
Also note that there is no full if statement ( if(..) {...}) in pcode, it's fust conditional jump labaled as if with goto altogether so it expects goto after every if.

**Comment by GhidorahRex** [comment]
> Also note that there is no full if statement ( if(..) {...}) in pcode, it's fust conditional jump labaled as if with goto altogether so it expects goto after every if.

This is true. If you have an `if ... else ...` that you need in Sleigh, then you would do:
```
if !(condition) goto <label>
  {do stuff}
  goto <end>
<label>
  {do other stuff}
<end>
```

But if you have an if/else to pick which of two things to assign to a variable, then you can use the one-line construct posted by @emteere above:
`result = (zext(cond) * valueA) + (zext(!cond) * valueB);`
We call it a choosery pattern. It's handled by the compiler as basically an if/else construct and also shortcuts that to just be a simple assignment if the condition is always true or always false.

**Comment by philpem** [comment]
The other thing worthy of note is what happens to the result -- the T flag affects that too. I might have missed the assignment statement in my T=1 example -- with T=1 the result is written back to the memory location, else the accumulator is used.

It seems like the if-else statement might be a better way to go.
